compiler/src/main/kotlin/se/ansman/kotshi/AdaptersProcessingStep.kt

```diff
@@ -63,501 +77,470 @@ class AdaptersProcessingStep(
         nameAllocator.newName("stringBuilder")
-        (0 until adapterKeys.size).forEach { nameAllocator.newName(generateAdapterFieldName(it)) }
+        nameAllocator.newName("it")
 
-        val jsonNames = properties
-            .filterNot { it.isTransient }
-            .map { it.jsonName }
-        val stringArguments = Collections.nCopies(jsonNames.size, "\$S").joinToString(",\n")
-        val optionsField = FieldSpec.builder(JsonReader.Options::class.java, "OPTIONS", Modifier.FINAL, Modifier.STATIC, Modifier.PRIVATE)
-            .initializer("\$[\$T.of(\n$stringArguments)\$]", JsonReader.Options::class.java, *jsonNames.toTypedArray())
-            .build()
-        val typeSpec = TypeSpec.classBuilder(adapter)
-            .addTypeVariables(genericTypes)
-            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
-            .superclass(getAdapterType(NamedJsonAdapter::class.java, typeName))
-            .applyIf(jsonNames.isNotEmpty()) {
-                addField(optionsField)
-            }
-            .addFields(generateFields(adapterKeys))
-            .addMethod(generateConstructor(element, typeElement, adapterKeys, genericTypes))
-            .addMethod(generateWriteMethod(typeMirror, properties, adapterKeys))
-            .addMethod(generateReadMethod(nameAllocator, typeMirror, properties, adapterKeys, optionsField))
-            .maybeAddGeneratedAnnotation(elements, sourceVersion)
+        val moshiParameter = ParameterSpec.builder("moshi", Moshi::class.java)
             .build()
 
-        val output = JavaFile.builder(adapter.packageName(), typeSpec).build()
+        val typesParameter = ParameterSpec.builder("types", Array<Type>::class.plusParameter(Type::class))
+            .build()
 
-        output.writeTo(filer)
-        adapters[TypeName.get(element.asType())] = when {
-            adapterKeys.isEmpty() -> GeneratedAdapter(adapter, requiresMoshi = false)
-            genericTypes.isNotEmpty() -> GeneratedAdapter(adapter, requiresTypes = true)
-            else -> GeneratedAdapter(adapter)
-        }
-    }
+        val adapterKeys = properties
+            .asSequence()
+            .filter { it.shouldUseAdapter }
+            .map { it.adapterKey }
+            .generatePropertySpecs(enclosingClass, moshiParameter, typesParameter, nameAllocator, typeVariables, imports)
 
-    private fun findConstructor(element: Element): ExecutableElement {
-        val constructors = ElementFilter.constructorsIn(element.enclosedElements)
+        val jsonNames = properties
+            .asSequence()
+            .filterNot { it.isTransient }
+            .map { it.jsonName }
+            .toList()
 
-        if (constructors.isEmpty()) {
-            throw ProcessingError("No constructors found", element)
-        }
+        val stringArguments = jsonNames.asSequence().map { "%S" }.joinToString(",\n")
+        val options = PropertySpec.builder("options", JsonReader.Options::class, KModifier.PRIVATE)
```

Comment:
I'm debating whether this should be in the companion class or not. We'd get an extra method but multiple adapters would share the same instance. Perhaps it should be in the companion for generic classes since they are the only ones who would be created multiple times.